lib/spack/spack/relocate.py

```diff
@@ -231,23 +228,63 @@ def macho_make_paths_placeholder(rpaths, deps, idpath):
     return (new_rpaths, new_deps, new_idpath)
 
 
-def macho_replace_paths(old_dir, new_dir, rpaths, deps, idpath):
+def macho_match_rpaths(old_dir, rpaths, env_rpaths):
+    """
+    Try to match rpath directory names using partial hash
+    """
+    new_rpaths = list()
+    for rpath in rpaths:
+        new_rpath = rpath
+        if re.match(old_dir, rpath):
+            hash = rpath.split(os.sep)[-2].split('-')[-1][:7]
+            for env_rpath in env_rpaths:
+                if re.search(hash, env_rpath):
+                    path_parts = env_rpath.split(os.sep)[:-1]
+                    path_parts.append(rpath.split(os.sep)[-1])
+                    new_rpath = os.sep.join(path_parts)
+                    new_rpaths.append('%s' % new_rpath)
+                    break
+        else:
+            new_rpaths.append('%s' % new_rpath)
+
+    return new_rpaths
+
+
+def macho_replace_paths(old_dir, new_dir, rpaths, deps, idpath, env_rpaths):
     """
     Replace old_dir with new_dir in rpaths, deps and idpath
     and return replacements
     """
+    new_rpaths = list()
+    if len(rpaths) == len(env_rpaths):
+        new_rpaths = list(env_rpaths)
+    else:
+        for rpath in macho_match_rpaths(old_dir, rpaths, env_rpaths):
+            new_rpaths.append(rpath)
+
     new_idpath = None
     if idpath:
-        new_idpath = idpath.replace(old_dir, new_dir)
-    new_rpaths = list()
+        idname = os.path.basename(idpath)
+        for rpath in rpaths:
+            if re.search(rpath, idpath):
+                for nrpath in new_rpaths:
+                    new_id = '%s/%s' % (nrpath, idname)
```

Comment:
`os.path.join`

Also, could you explain how this is supposed to work in terms of the expected format of the rpaths and idpath? I'm not clear on it from reading the logic.

Reply:
I guess the id path would just be in the new prefix. I will update. I initially was replacing with @rpath/idname but that needed the ability to change the rpaths. When installing macOS libs on linux using macholib, there is no method to change the rpaths. I instead switched to find the dependecies using the rpaths and duplicated the search for the id path. I see not the search the prefix would be better.